agb/src/agb_alloc/block_allocator.rs

```diff
@@ -21,31 +26,79 @@ impl Block {
             aligned_to.align(),
         )
         .expect("too large allocation")
+        .align_to(8)
+        .expect("too large allocation")
     }
 }
 
 struct BlockAllocatorState {
-    first_free_block: Option<NonNull<Block>>,
+    first_free_block: Option<SendNonNull<Block>>,
 }
 
 pub(crate) struct BlockAllocator {
     inner_allocator: BumpAllocator,
-    state: Mutex<BlockAllocatorState>,
+    state: Mutex<RefCell<BlockAllocatorState>>,
 }
 
 impl BlockAllocator {
     pub(super) const unsafe fn new() -> Self {
         Self {
             inner_allocator: BumpAllocator::new(),
-            state: Mutex::new(BlockAllocatorState {
+            state: Mutex::new(RefCell::new(BlockAllocatorState {
                 first_free_block: None,
-            }),
+            })),
         }
     }
 
-    unsafe fn new_block(&self, layout: Layout) -> *mut u8 {
+    #[allow(dead_code)]
+    pub unsafe fn number_of_blocks(&self) -> u32 {
+        free(|key| {
+            let mut state = self.state.borrow(*key).borrow_mut();
+
+            let mut count = 0;
+
+            let mut list_ptr = &mut state.first_free_block;
+            while let Some(mut curr) = list_ptr {
+                count += 1;
+                list_ptr = &mut curr.as_mut().next;
+            }
+
+            count
+        })
+    }
+
+    fn new_block(&self, layout: Layout, cs: &CriticalSection) -> *mut u8 {
         let overall_layout = Block::either_layout(layout);
-        self.inner_allocator.alloc(overall_layout)
+        self.inner_allocator.alloc_critical(overall_layout, cs)
+    }
+
+    unsafe fn normalise(&self) {
+        free(|key| {
+            let mut state = self.state.borrow(*key).borrow_mut();
+
+            let mut list_ptr = &mut state.first_free_block;
+
+            while let Some(mut curr) = list_ptr {
+                if let Some(next_elem) = curr.as_mut().next {
+                    let difference = next_elem
+                        .as_ptr()
+                        .cast::<u8>()
+                        .offset_from(curr.as_ptr().cast::<u8>());
+                    let usize_difference: usize = difference
+                        .try_into()
+                        .expect("distances in alloc'd blocks must be positive");
+
+                    if usize_difference == curr.as_mut().size {
+                        let current = curr.as_mut();
+                        let next = next_elem.as_ref();
+
+                        current.size += next.size;
+                        current.next = next.next;
+                    }
```

Comment:
If I'm understanding correctly, if you have 3 in a row that need combining, you only combine the first two, then jump to the next one?

Feels like you need a `continue` on the line after `current.next = next.next` to check if there is yet another to combine?